mitodl/open-discussions
# Added Elasticsearch document and added indexing handlers for posts

#### What are the relevant tickets?
Closes #605

#### What's this PR do?
- Adds the elasticsearch document mapping, as well as a function that clears and creates the index in ES
- Wraps various channel API methods to make changes to the index when there are changes made to the information that we index for posts (and only posts)

#### How should this be manually tested?
- Enable the `INDEX_UPDATES` feature (i.e.: add `FEATURE_INDEX_UPDATES=true` to your `.env`)
- Run `search.indexing_api.clear_and_create_index` in a Django shell
- Create posts and alter posts in different ways (changing text, adding comments, upvoting/downvoting)
- Query the index and confirm that those changes are reflected in the indexed document (or that the document was created if it's a new post)

An easy curl command to inspect the documents in the index:
```bash
curl -XGET -s 'http://localhost:9101/discussions_local/_search?pretty=true&filter_path=hits.hits._id,hits.hits._source&sort=created:desc'
```

#### Where should the reviewer start?
Probably `channels/api.py`

#### Any background context you want to provide?
- This PR does not add a search UI or add any querying capability. That can and should be done in a separate issue.
- This PR does not index changes to comments. That can also be done in a separate PR.
- `search/connection.py` was copied from Micromasters. If we want to make any changes to that connection logic we can, but worth pointing out that it's copied
